src/G2/Equiv/Induction.hs

```diff
@@ -0,0 +1,336 @@
+replaceScrutinee e1 e2 (Case e i a) = Case (replaceScrutinee e1 e2 e) i a
+replaceScrutinee _ _ e = e
+
+-- checking for depth of first expression within second
+scrutineeDepth :: Expr -> Expr -> Int
+scrutineeDepth e1 e2 | e1 == e2 = 0
+scrutineeDepth e1 (Tick _ e) = scrutineeDepth e1 e
+scrutineeDepth e1 (Case e i a) = 1 + scrutineeDepth e1 e
+scrutineeDepth _ _ = error "Not Contained"
+
+-- the depths do not need to be the same
+-- however, the stamps should match up to the depth of the old one
+validScrutinee :: StateET -> StateET -> StateET -> Bool
+validScrutinee s p pc =
+  let d = scrutineeDepth (exprExtract p) (exprExtract pc)
+      stamps_old = take d $ readStamps (exprExtract pc)
+      stamps_new = take d $ readStamps (exprExtract s)
+  in trace (show (d, stamps_old, stamps_new)) stamps_old == stamps_new
+
+innerScrutineeStates :: State t -> [State t]
+innerScrutineeStates s@(State { curr_expr = CurrExpr _ e }) =
+  map (\e' -> s { curr_expr = CurrExpr Evaluate e' }) (innerScrutinees e)
+
+-- TODO keep track of the whole expression that was used for the inner scrutinee
+moreRestrictiveIndRight :: S.Solver solver =>
+                           solver ->
+                           HS.HashSet Name ->
+                           [(StateET, StateET)] ->
+                           (StateET, StateET) ->
+                           W.WriterT [Marker] IO (Maybe (PrevMatch EquivTracker))
+moreRestrictiveIndRight solver ns prev (s1, s2) =
+  let prev1 = map (\(p1, p2) -> (p1, p2, innerScrutineeStates p2)) prev
+      prev2 = [(p1, p2', p2) | (p1, p2, p2l) <- prev1, p2' <- p2l]
+  in moreRestrictivePairAux solver ns prev2 (s1, s2)
+
+-- substitution happens on the left here
+-- TODO now the right-hand state returned would always be s2
+-- TODO collect information about the expressions used for induction
+-- TODO return a partially-defined IndMarker?
+inductionL :: S.Solver solver =>
+              solver ->
+              HS.HashSet Name ->
+              [(StateET, StateET)] ->
+              (StateET, StateET) ->
+              W.WriterT [Marker] IO (Bool, StateET, IndMarker)
+inductionL solver ns prev (s1, s2) = do
+  let scr1 = innerScrutinees $ exprExtract s1
+      scr2 = innerScrutinees $ exprExtract s2
+      scr_pairs = [(sc1, sc2) | sc1 <- scr1, sc2 <- scr2]
+      scr_states = [(s1 { curr_expr = CurrExpr Evaluate sc1 }, s2 { curr_expr = CurrExpr Evaluate sc2 }) | (sc1, sc2) <- scr_pairs]
+  mr_pairs <- mapM (moreRestrictiveIndRight solver ns prev) scr_states
+  let mr_zipped = zip scr_pairs mr_pairs
+      working_info = [(sc1, sc2, pm) | ((sc1, sc2), Just pm) <- mr_zipped]
+      working_info' = filter (\(_, _, PrevMatch _ (_, p2) _ pc2) -> validScrutinee s2 p2 pc2) working_info
+  case working_info' of
+    [] -> return (False, s1, undefined)
+    h:_ -> let (sc1, sc2, PrevMatch (q1, q2) (p1, p2) (mapping, _) pc2) = h
+               e2_old = exprExtract pc2
+               hm_list = HM.toList mapping
```

Comment:
What does this represent?  The name hm_list is not very informative.  Is it the mapping between variables in one state to expressions in another state?